pkg/controller/controller_instance.go

```diff
@@ -370,81 +389,133 @@ func (c *controller) pollInstance(serviceClass *v1alpha1.ServiceClass, servicePl
 	}
-	resp, rc, err := brokerClient.PollServiceInstance(instance.Spec.ExternalID, lastOperationRequest)
+
+	glog.V(5).Infof("Polling last operation on Instance %v/%v", instance.Namespace, instance.Name)
+
+	response, err := brokerClient.PollLastOperation(request)
 	if err != nil {
-		glog.Warningf("Poll failed for %v/%v  : %s", instance.Namespace, instance.Name, err)
-		return err
-	}
-	glog.V(4).Infof("Poll for %v/%v returned %q : %q", instance.Namespace, instance.Name, resp.State, resp.Description)
+		// If the operation was for delete and we receive a http.StatusGone,
+		// this is considered a success as per the spec, so mark as deleted
+		// and remove any finalizers.
+		if osb.IsGoneError(err) && deleting {
+			clone, err := api.Scheme.DeepCopy(instance)
+			if err != nil {
+				return err
+			}
+			toUpdate := clone.(*v1alpha1.Instance)
 
-	// If the operation was for delete and we receive a http.StatusGone,
-	// this is considered a success as per the spec, so mark as deleted
-	// and remove any finalizers.
-	if rc == http.StatusGone && deleting {
-		instance.Status.AsyncOpInProgress = false
-		// Clear the finalizer
-		if finalizers := sets.NewString(instance.Finalizers...); finalizers.Has(v1alpha1.FinalizerServiceCatalog) {
-			finalizers.Delete(v1alpha1.FinalizerServiceCatalog)
-			c.updateInstanceFinalizers(instance, finalizers.List())
+			toUpdate.Status.AsyncOpInProgress = false
+			c.updateInstanceCondition(
+				toUpdate,
+				v1alpha1.InstanceConditionReady,
+				v1alpha1.ConditionFalse,
+				successDeprovisionReason,
+				successDeprovisionMessage,
+			)
+
+			// Clear the finalizer
+			if finalizers := sets.NewString(toUpdate.Finalizers...); finalizers.Has(v1alpha1.FinalizerServiceCatalog) {
+				finalizers.Delete(v1alpha1.FinalizerServiceCatalog)
+				c.updateInstanceFinalizers(toUpdate, finalizers.List())
+			}
+
+			c.recorder.Event(instance, api.EventTypeNormal, successDeprovisionReason, successDeprovisionMessage)
+			glog.V(5).Infof("Successfully deprovisioned Instance %v/%v of ServiceClass %v at Broker %v", instance.Namespace, instance.Name, serviceClass.Name, brokerName)
+			return nil
 		}
-		c.updateInstanceCondition(
-			instance,
-			v1alpha1.InstanceConditionReady,
-			v1alpha1.ConditionFalse,
-			successDeprovisionReason,
-			successDeprovisionMessage,
-		)
-		c.recorder.Event(instance, api.EventTypeNormal, successDeprovisionReason, successDeprovisionMessage)
-		glog.V(5).Infof("Successfully deprovisioned Instance %v/%v of ServiceClass %v at Broker %v", instance.Namespace, instance.Name, serviceClass.Name, brokerName)
-		return nil
+
+		// We got some kind of error from the broker.  While polling last
+		// operation, this represents an invalid response and we should
+		// continue polling last operation.
+		//
+		// The ready condition on the instance should already have
+		// condition false; it should be sufficient to create an event for
+		// the instance.
+		errText := ""
+		if osb.IsHTTPError(err) {
```

Comment:
Yeah, we are definitely approaching the point where it will become a necessity to apply extract method and break stuff up.  I was pretty militant about _not_ doing that when a lot of people in the group were learning the controller pattern.  The reason for that is that extract-method application tends to solidify a design in a lot of people's heads and shape how they think about a problem space.  I've seen that go very badly in other parts of kubernetes (and in other projects in my career), and the way we remedied it was by refactoring to this exploded style of factoring which is quite long-winded, but fairly clear in the sense of what the code is doing.

Basically, I wanted to avoid some mistake in applying extract-method result in the mentality of 'we can't do it X way, we'd have to change all this code in the controller'!  Having code that literally no one is happy with helps people developing mental barricades against a design that might be better because they've fallen in love with a refactor PR they did.

If we can possibly, possibly hang on, I'd prefer to leave it this way (the mental handle I have for this style of code is 'unfactored') until we have the OSB integration completed (ie, plan update, param update).  I'm not sure if that will be possible, though.

LMK if this comment makes any sense.